rfd/0228-resource-scoped-constraints-in-access-requests.md

```diff
@@ -0,0 +1,168 @@
+- **SSH node (login + optional sudo)**:  
+  Diego can SSH to a service host as `deploy` but needs a temporary `admin` login (and possibly `sudo`) to perform maintenance. In Teleport's Web UI, he clicks the connect dropdown for the service node, and sees `deploy` under his **granted** roles, and `admin` (with an "allow sudo" checkbox) under the **requestable** section. He selects `admin`, checks "allow sudo", submits the request, and after approval, connects with just `admin` (and the approved sudo constraint).
+- **Database (role/user)**:  
+  Chandra has read-only access to a Postgres DB as `report_reader` and needs `migration_admin` for a one-off schema change. In Teleport's Web UI, she clicks the connect dropdown for the Postgres app, selects `migration_admin` under the **requestable** section, submits the request, and after approval, connects with only the `migration_admin` DB role for the session.
+
+**Reviewer experience**: Reviewers see both **(a)** the requested constraints and **b)** the resolved roles computed to satisfy them.
+
+## Design Overview
+
+### ResourceID & Constraints
+
+> Context: `ResourceID` identifies resources in Access Requests and is also used in certs (`AllowedResourceIDs`) to scope post-approval authorization.
+
+We add an optional, typed `Constraints` field to`ResourceID`. The content is domain-specific and enumerated.
+
+#### Go
+
+```go
+package types
+
+type ResourceID struct {
+  // 1..4 are existing fields.
+  Kind            string              `json:"kind"`
+  Name            string              `json:"name"`
+  ClusterName     string              `json:"clusterName"`
+  SubResourceName string              `json:"subResourceName,omitempty"`
+  // New: Optional constraints for this resource.
+  Constraints     *RequestConstraints `json:"constraints,omitempty"`
+}
+
+type ConstraintDomain int32
+
+const (
+  ConstraintDomainUnspecified ConstraintDomain = iota
+  ConstraintDomainAWSConsole
+  ConstraintDomainAWSIdentityCenter
+  ConstraintDomainSSH
+  ConstraintDomainDatabase
+)
+
+type RequestConstraints struct {
+  Domain ConstraintDomain   `json:"domain"`
+
+  // Exactly one of the following non-nil, based on Domain.
+  AWS *AWSConstraints `json:"aws,omitempty"`
+  SSH *SSHConstraints `json:"ssh,omitempty"`
+  DB  *DBConstraints  `json:"db,omitempty"`
+}
+
+type AWSConstraints struct {
+  RoleARNs       []string `json:"role_arns,omitempty"`        // AWS Console
+  PermissionSets []string `json:"permission_sets,omitempty"`  // AWS IC
+}
+
+type SSHConstraints struct {
+  Logins    []string `json:"logins,omitempty"`
+  AllowSudo *bool    `json:"allow_sudo"`                      // example non-principal constraint
```

Comment:
Let's remove this from scope for now, don't know what this would mean yet. Let's focus on principals.
```suggestion
```

Reply:
Tried to clarify it's in here for example of future uses but can remove if it's just distracting. Should we also keep out the user stories for ssh/db uses as well and just focus on AWS for now?